File: src/main/java/org/c3lang/intellij/annotation/CallExpr.java
package org.c3lang.intellij.annotation;

import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import org.c3lang.intellij.C3Language;
import org.c3lang.intellij.psi.C3CallExpr;
import org.c3lang.intellij.psi.C3File;
import org.c3lang.intellij.psi.C3FuncDefinition;
import org.c3lang.intellij.psi.C3MacroDefinition;
import org.c3lang.intellij.psi.C3ModuleDefinition;
import org.c3lang.intellij.psi.C3ModuleSection;
import org.c3lang.intellij.psi.C3TopLevel;
import org.c3lang.intellij.psi.ModuleName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;

public final class CallExpr
{
    private CallExpr()
    {
    }

    public static void annotateCallExpr(@NotNull C3CallExpr element, @NotNull AnnotationHolder holder)
    {
        if (!(element.getContainingFile() instanceof C3File file)) return;

        String text = element.getText();
        int callStart = text.indexOf('(');
        if (callStart < 0) return;

        String fullName = text.substring(0, callStart);
        String module = "";
        ModuleName moduleName = element.getModuleDefinition().getModuleName();
        if (moduleName != null)
        {
            module = moduleName.getValue();
        }
        int moduleSeparator = fullName.lastIndexOf("::");
        if (moduleSeparator >= 0)
        {
            module = fullName.substring(0, moduleSeparator);
        }
        String name = moduleSeparator >= 0 ? fullName.substring(moduleSeparator + 2) : fullName;

        PsiElement declaration = findDeclaration(element.getProject(), element, file, module, name, holder);
        if (declaration == null)
        {
            holder.newAnnotation(HighlightSeverity.WARNING, "Declaration not found").create();
        }
    }

    @Nullable
    private static PsiElement findDeclaration(
            @NotNull Project project,
            @NotNull C3CallExpr element,
            @NotNull C3File file,
            @NotNull String module,
            @NotNull String name,
            @NotNull AnnotationHolder holder)
    {
        if (module.isEmpty())
        {
            ArrayList<PsiElement> declarations = findDeclarationInFile(file, name);
            return declarations.isEmpty() ? null : declarations.get(0);
        }

        ArrayList<PsiElement> declarations = findDeclarationInModule(project, module, name);
        if (declarations.isEmpty()) return null;

        PsiElement firstMatch = declarations.get(0);
        C3ModuleDefinition moduleDefinition = element.getModuleDefinition();
        ModuleName matchModule = null;
        if (firstMatch instanceof C3FuncDefinition definition)
        {
            matchModule = definition.getModuleDefinition().getModuleName();
        }
        else if (firstMatch instanceof C3MacroDefinition definition)
        {
            matchModule = definition.getModuleDefinition().getModuleName();
        }

        boolean hasValidImport = false;
        if (matchModule != null)
        {
            for (ModuleName imported : moduleDefinition.getImports())
            {
                if (imported.getValue().equals(matchModule.getValue()))
                {
                    hasValidImport = true;
                    break;
                }
            }
        }

        if (!hasValidImport)
        {
//            holder.newAnnotation(HighlightSeverity.WARNING, "Missing import")
//                    .range(element)
//                    .newLocalQuickFix(new AddMissingImportFix(), InspectionManager.getInstance(project).createProblemDescriptor("", new AddMissingImportFix()))
//                    .registerFix();
        }

        return firstMatch;
    }

    @NotNull
    public static ArrayList<PsiElement> findDeclarationInModule(
            @NotNull Project project,
            @NotNull String module,
            @NotNull String name)
    {
        File stdLibPath = guessStdLibPath();
        ArrayList<PsiElement> matches = new ArrayList<>();
        if (stdLibPath != null)
        {
            for (C3File file : walkStdLib(project, stdLibPath))
            {
                addModuleMatches(file, module, name, matches);
            }
        }

        PsiManager psiManager = PsiManager.getInstance(project);
        VirtualFile[] contentRoots = ProjectRootManager.getInstance(project).getContentRoots();
        for (VirtualFile root : contentRoots)
        {
            VfsUtilCore.iterateChildrenRecursively(root, null, file -> {
                if (!file.isDirectory())
                {
                    PsiFile psiFile = psiManager.findFile(file);
                    if (psiFile instanceof C3File c3File && psiFile.getLanguage() == C3Language.INSTANCE)
                    {
                        addModuleMatches(c3File, module, name, matches);
                    }
                }
                return true;
            });
        }

        return matches;
    }

    @NotNull
    public static ArrayList<C3File> walkStdLib(@NotNull Project project, @NotNull File stdLibPath)
    {
        VirtualFile virtualRoot = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(stdLibPath);
        ArrayList<C3File> files = new ArrayList<>();
        if (virtualRoot == null) return files;

        PsiManager psiManager = PsiManager.getInstance(project);
        walkStdLib(virtualRoot, psiManager, files);
        return files;
    }

    private static void walkStdLib(
            @NotNull VirtualFile virtualFile,
            @NotNull PsiManager psiManager,
            @NotNull List<C3File> files)
    {
        if (virtualFile.isDirectory())
        {
            for (VirtualFile child : virtualFile.getChildren())
            {
                walkStdLib(child, psiManager, files);
            }
            return;
        }

        PsiFile psi = psiManager.findFile(virtualFile);
        if (psi instanceof C3File c3File && psi.getLanguage() == C3Language.INSTANCE)
        {
            files.add(c3File);
        }
    }

    @Nullable
    public static File guessStdLibPath()
    {
        return guessStdLibPath(new File("/"));
    }

    @Nullable
    public static File guessStdLibPath(@NotNull File startDir)
    {
        ArrayDeque<File> queue = new ArrayDeque<>();
        queue.add(startDir);

        while (!queue.isEmpty())
        {
            File current = queue.removeFirst();
            if (!current.isDirectory()) continue;

            File lib = new File(current, "lib");
            File std = new File(lib, "std");
            if ("c3c".equals(current.getName()) && lib.exists() && std.exists() && std.isDirectory())
            {
                return std;
            }

            File[] children = current.listFiles(File::isDirectory);
            if (children != null)
            {
                for (File child : children)
                {
                    queue.add(child);
                }
            }
        }

        return null;
    }

    @NotNull
    public static ArrayList<PsiElement> findDeclarationInFile(@NotNull C3File file, @NotNull String name)
    {
        ArrayList<PsiElement> matches = new ArrayList<>();
        for (PsiElement child : file.getChildren())
        {
            if (child instanceof C3ModuleSection moduleSection)
            {
                addTopLevelMatches(moduleSection.getTopLevelList(), name, matches);
            }
        }
        return matches;
    }

    private static void addModuleMatches(
            @NotNull C3File file,
            @NotNull String module,
            @NotNull String name,
            @NotNull List<PsiElement> matches)
    {
        for (PsiElement child : file.getChildren())
        {
            if (!(child instanceof C3ModuleSection moduleSection)) continue;

            ModuleName moduleName = moduleSection.getModuleName();
            if (moduleName != null && moduleName.getValue().endsWith(module))
            {
                addTopLevelMatches(moduleSection.getTopLevelList(), name, matches);
            }
        }
    }

    private static void addTopLevelMatches(
            @NotNull List<C3TopLevel> topLevels,
            @NotNull String name,
            @NotNull List<PsiElement> matches)
    {
        for (C3TopLevel topLevel : topLevels)
        {
            C3FuncDefinition function = topLevel.getFuncDefinition();
            if (function != null && function.getFuncDef().getFuncHeader().getFuncName().getText().equals(name))
            {
                matches.add(function);
            }

            C3MacroDefinition macro = topLevel.getMacroDefinition();
            if (macro != null && macro.getMacroHeader().getMacroName().getText().equals(name))
            {
                matches.add(macro);
            }
        }
    }
}

File: src/main/java/org/c3lang/intellij/annotation/CallExpr.kt
//package org.c3lang.intellij.annotation
//
//import com.intellij.lang.annotation.AnnotationHolder
//import com.intellij.lang.annotation.HighlightSeverity
//import com.intellij.openapi.project.Project
//import com.intellij.openapi.roots.ProjectRootManager
//import com.intellij.openapi.vfs.LocalFileSystem
//import com.intellij.openapi.vfs.VfsUtilCore
//import com.intellij.openapi.vfs.VirtualFile
//import com.intellij.psi.PsiElement
//import com.intellij.psi.PsiManager
//import org.c3lang.intellij.C3Language
//import org.c3lang.intellij.psi.C3CallExpr
//import org.c3lang.intellij.psi.C3File
//import org.c3lang.intellij.psi.C3FuncDefinition
//import org.c3lang.intellij.psi.C3MacroDefinition
//import org.c3lang.intellij.psi.C3ModuleSection
//import org.c3lang.intellij.psi.C3TopLevel
//import java.io.File
//
//// TODO: add quick fix
//// TODO: check if import is present in case of other module and prompt to add it
//internal fun annotateCallExpr(element: C3CallExpr, holder: AnnotationHolder)
//{
//    val file = element.containingFile as C3File
//    val fullName = element.text.substring(0, element.text.indexOfFirst { it == '(' })
//    var module = element.moduleDefinition.moduleName?.value ?: ""
//    if (fullName.contains("::"))
//    {
//        module = fullName.substring(0, fullName.lastIndexOf("::"))
//    }
//    val name = fullName.substringAfterLast("::")
//
//    // TODO: add sout and soutf shortcuts
//
//    // TODO: if module is empty set module to current module
//    // TODO: if current module is default module search current file
//
//    /**
//     * @see C3FuncDefinition
//     * @see C3MacroDefinition
//     */
//    val declaration = findDeclaration(file.project, element, file, module, name, holder)
//
//    if (declaration == null)
//    {
//        holder.newAnnotation(HighlightSeverity.WARNING, "Declaration not found")
//            .create()
//    }
//}
//
//// TODO: scan all files in project and std lib for modules with matching name
//// TODO: then search that module for functions with the same name
//private fun findDeclaration(project: Project, element: C3CallExpr, file: C3File, module: String, name: String, holder: AnnotationHolder): PsiElement?
//{
//    if (module.isEmpty())
//    {
//        val decl = findDeclarationInFile(file, name)
//        return if (decl.isNotEmpty())
//        {
//            decl.first()
//        } else
//        {
//            null
//        }
//    } else
//    {
//        val decl = findDeclarationInModule(file.project, module, name)
//
//        if (decl.isNotEmpty())
//        {
//            val firstMatch = decl.first()
//            val validImports = arrayListOf<String>()
//
//            if (firstMatch is C3MacroDefinition)
//            {
//                validImports.addAll(element.moduleDefinition.imports.filter { it.value == firstMatch.moduleDefinition.moduleName?.value }.map { it.value })
//            } else if (firstMatch is C3FuncDefinition)
//            {
//                validImports.addAll(element.moduleDefinition.imports.filter { it.value == firstMatch.moduleDefinition.moduleName?.value }.map { it.value })
//            }
//
//            if (validImports.isEmpty())
//            {
////                holder.newAnnotation(HighlightSeverity.WARNING, "Missing import")
////                    .range(element)
////                    .newLocalQuickFix(AddMissingImportFix(), InspectionManager.getInstance(project).createProblemDescriptor("", AddMissingImportFix()) as ProblemDescriptor)
////                    .registerFix()
//            }
//
//            return firstMatch
//        } else
//        {
//            return null
//        }
//    }
//}
//
//fun findDeclarationInModule(project: Project, module: String, name: String): ArrayList<PsiElement>
//{
//    val stdLibPath = guessStdLibPath() ?: return arrayListOf()
//    val stdLibFiles = walkStdLib(project, stdLibPath)
//    val matches = arrayListOf<PsiElement>()
//
//    stdLibFiles.forEach { file ->
//        val modules = file.children.filterIsInstance<C3ModuleSection>().filter { it.moduleName?.value?.endsWith(module) ?: false }
//
//        modules.forEach { module ->
//            val topLevels = module.children.filterIsInstance<C3TopLevel>()
//
//            topLevels.forEach { topLevel ->
//                topLevel.children.filterIsInstance<C3FuncDefinition>().filter { it.funcDef.funcHeader.funcName.text == name }.forEach { matches.add(it) }
//                topLevel.children.filterIsInstance<C3MacroDefinition>().filter { it.macroHeader.macroName.text == name }.forEach { matches.add(it) }
//            }
//        }
//    }
//
//    val projectFiles = mutableListOf<C3File>()
//    val psiManager = PsiManager.getInstance(project)
//
//    val contentRoots = ProjectRootManager.getInstance(project).contentRoots
//    for (root in contentRoots)
//    {
//        VfsUtilCore.iterateChildrenRecursively(root, null) { file ->
//            if (!file.isDirectory)
//            {
//                psiManager.findFile(file)?.let { psiFile ->
//                    if (psiFile.language == C3Language.INSTANCE)
//                    {
//                        projectFiles.add(psiFile as C3File)
//                    }
//                }
//            }
//            true
//        }
//    }
//
//    // TODO: add module search
//    projectFiles.forEach { file ->
//        val modules = file.children.filterIsInstance<C3ModuleSection>().filter { it.moduleName?.value?.endsWith(module) ?: false }
//
//        modules.forEach { module ->
//            val topLevels = module.children.filterIsInstance<C3TopLevel>()
//
//            topLevels.forEach { topLevel ->
//                topLevel.children.filterIsInstance<C3FuncDefinition>().filter { it.funcDef.funcHeader.funcName.text == name }.forEach { matches.add(it) }
//                topLevel.children.filterIsInstance<C3MacroDefinition>().filter { it.macroHeader.macroName.text == name }.forEach { matches.add(it) }
//            }
//        }
//    }
//
//    return matches
//}
//
//fun walkStdLib(project: Project, stdLibPath: File): ArrayList<C3File>
//{
//    val virtualRoot = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(stdLibPath) ?: return arrayListOf()
//    val psiManager = PsiManager.getInstance(project)
//    val files = arrayListOf<C3File>()
//
//    fun recurse(vf: VirtualFile)
//    {
//        if (vf.isDirectory)
//        {
//            vf.children.forEach { recurse(it) }
//        } else
//        {
//            val psi = psiManager.findFile(vf)
//            if (psi != null && psi.language == C3Language.INSTANCE)
//            {
//                files.add(psi as C3File)
//            }
//        }
//    }
//
//    recurse(virtualRoot)
//
//    return files
//}
//
///**
// * Expensive operation
// *
// * @param startDir the start directory to start looking from
// * @return the guessed library directory
// */
//fun guessStdLibPath(startDir: File = File("/")): File?
//{
//    val queue = ArrayDeque<File>()
//    queue.add(startDir)
//
//    while (queue.isNotEmpty())
//    {
//        val current = queue.removeFirst()
//
//        if (current.isDirectory)
//        {
//            val lib = File(current, "lib")
//            val std = File(lib, "std")
//            if (current.name == "c3c" && lib.exists() && std.exists() && std.isDirectory)
//            {
//                return std
//            }
//
//            current.listFiles()?.filter { it.isDirectory }?.let { queue.addAll(it) }
//        }
//    }
//
//    return null
//}
//
//fun findDeclarationInFile(file: C3File, name: String): ArrayList<PsiElement>
//{
//    // TODO: find module -> find top level -> find function, macro
//    val modules = file.children.filterIsInstance<C3ModuleSection>()
//    val matches = arrayListOf<PsiElement>()
//
//    modules.forEach { module ->
//        val topLevels = module.children.filterIsInstance<C3TopLevel>()
//        topLevels.forEach { topLevel ->
//            topLevel.children.filterIsInstance<C3FuncDefinition>().filter { it.funcDef.funcHeader.funcName.text == name }.forEach { matches.add(it) }
//            topLevel.children.filterIsInstance<C3MacroDefinition>().filter { it.macroHeader.macroName.text == name }.forEach { matches.add(it) }
//        }
//    }
//
//    return matches
////    return file.children.filterIsInstance<C3FuncDefinition>().first { it.funcDef.funcHeader.funcName.text == name }
//}